go/arrow/cdata/cdata_exports.go

```diff
@@ -368,34 +368,37 @@ func exportArray(arr arrow.Array, out *CArrowArray, outSchema *CArrowSchema) {
 		exportField(arrow.Field{Type: arr.DataType()}, outSchema)
 	}
 
+	nbuffers := len(arr.Data().Buffers())
+	buf_offset := 0
+	// Some types don't have validity bitmaps, but we keep them shifted
+	// to make processing easier in other contexts. This means that
+	// we have to adjust when exporting.
+	has_validity_bitmap := internal.DefaultHasValidityBitmap(arr.DataType().ID())
+	if nbuffers > 0 && !has_validity_bitmap {
+		nbuffers--
+		buf_offset++
+	}
+
 	out.dictionary = nil
 	out.null_count = C.int64_t(arr.NullN())
 	out.length = C.int64_t(arr.Len())
 	out.offset = C.int64_t(arr.Data().Offset())
-	out.n_buffers = C.int64_t(len(arr.Data().Buffers()))
-
-	if out.n_buffers > 0 {
-		var (
-			nbuffers = len(arr.Data().Buffers())
-			bufs     = arr.Data().Buffers()
-		)
-		// unions don't have validity bitmaps, but we keep them shifted
-		// to make processing easier in other contexts. This means that
-		// we have to adjust for union arrays
-		if !internal.DefaultHasValidityBitmap(arr.DataType().ID()) {
-			out.n_buffers--
-			nbuffers--
-			bufs = bufs[1:]
-		}
+	out.n_buffers = C.int64_t(nbuffers)
+	out.buffers = nil
+
+	if nbuffers > 0 {
+		bufs := arr.Data().Buffers()
 		buffers := allocateBufferPtrArr(nbuffers)
-		for i := range bufs {
-			buf := bufs[i]
+		for i := 0; i < nbuffers; i++ {
```

Comment:
You could do:

```go
for i, buf := range buf[buf_offset:] {
```

`i` should still go from 0 -> `nbuffers` since `nbuffers` should be equal to `len(buf[buf_offset:])`

Alternately if we want to make sure we're being more explicit you could do:

```go
for i, buf := range bufs[buf_offset:buf_offset+nbuffers] {
```
But i don't think it's necessary.

Reply:
Hmm, but then the `i > 0` test below would be a bit more awkward to spell out.

Reply:
how so? The situations should be equivalent

if `len(bufs)` ==3, `buf_offset` == 1 and `nbuffers` == 2, your iterations would be:

1. `i` == 0, `buf` == `bufs[1]`
2. `i` == 1, `buf` == `bufs[2]`

which would be the same regardless of if you use `for i := 0; i < nbuffers; i++` or `for i, buf := range bufs[buf_offset:]`

right? or am i missing something?

Reply:
Oh, yes, sorry. It appears I had misread your proposal.